Updates to Teradata Provider

sc250072 opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->

Implemented enhancements to the Teradata Provider for Teradata.

This release incorporates the following features:

• Introduction of Stored Procedure Support in Teradata Hook
• Inclusion of the TeradataStoredProcedureOperator for executing stored procedures
• Integration of Azure Blob Storage to Teradata Transfer Operator
• Integration of Amazon S3 to Teradata Transfer Operator
• Provision of necessary documentation, along with unit and system tests, for the Teradata Provider modifications.

Teradata Provider System Health Dashboard: https://teradata.github.io/airflow/

Teradata Provider documentation build status: https://github.com/Teradata/airflow/actions/workflows/ci-teradata-documentation.yml 



---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

potiuk reviewed (approved):
Nice set of updates !

eladkal commented on airflow/providers/teradata/transfers/azure_blob_to_teradata.py:
```diff
@@ -0,0 +1,105 @@
+                        SELECT * FROM (
+                            LOCATION = '{self.blob_source_key}'
+                            ACCESS_ID= '{access_id}'
+                            ACCESS_KEY= '{access_secret}'
```
Won't this expose the secret in the logs?

sc250072 replied:
![Access_Id_Key_Log](https://github.com/apache/airflow/assets/23031170/a28a4285-0799-4485-8993-0fc9b9478414)
secret is logging as masked in the log.

eladkal commented on airflow/providers/teradata/transfers/azure_blob_to_teradata.py:
```diff
@@ -0,0 +1,105 @@
+                    "There is an issue with the transfer operation. Please validate azure and "
+                    "teradata connection details."
+                )
+                raise
```
I'm curious. How are the original error message look like?
Does these messages added more value on top of the original error message?
i would expect the service to return all the needed information in the exception itself

sc250072 replied:
Agreed and removed additional checking of exceptions instead added log error for exception received. Teradata returning needed exception information.

satish-chinthanippu committed:
Updates to Teradata Provider with release 2.2.0

Implemented enhancements to the Teradata Provider for Teradata.

This release incorporates the following features:

• Introduction of Stored Procedure Support in Teradata Hook
• Inclusion of the TeradataStoredProcedureOperator for executing stored procedures
• Integration of Azure Blob Storage to Teradata Transfer Operator
• Integration of Amazon S3 to Teradata Transfer Operator
• Provision of necessary documentation, along with unit and system tests, for the Teradata Provider modifications.

satish-chinthanippu committed:
PR initial review comments addressed (#36)

satish-chinthanippu committed:
Merge branch 'apache:main' into main

eladkal commented on airflow/providers/teradata/provider.yaml:
```diff
@@ -33,6 +33,8 @@ dependencies:
   - teradatasqlalchemy>=17.20.0.0
   - teradatasql>=17.20.0.28
+  - apache-airflow-providers-microsoft-azure
+  - apache-airflow-providers-amazon
```
Setting this means that anyone who is using teradata provider will be forced to also have azure and amazon.
I don't think it's right as these integrations are not part of the core functionality that this provider offer (unlike for example common-sql) I suggest to move them to optional dependencies like:

https://github.com/apache/airflow/blob/1074b8eed680af9668f11c63cf28e72db5470fde/airflow/providers/google/provider.yaml#L167-L185

This will allow users of teradata to choose if they want to add the optional dependencies into their installation.
You can also wrap the imports with AirflowOptionalProviderFeatureException to prevent cases where someone uses operator without having the underlying provider installed (similar to what I suggested in https://github.com/apache/airflow/pull/39366#discussion_r1587563547 )

sc250072 replied:
@eladkal  thank you. Will change it.

sc250072 replied:
Modified as suggested

satish-chinthanippu committed:
Generated provider dependenices

eladkal commented on airflow/providers/teradata/hooks/teradata.py:
```diff
@@ -32,6 +32,17 @@
+PARAM_TYPES = {bool, float, int, str}
+
+
+def _map_param(value):
+    if value in PARAM_TYPES:
+        # In this branch, value is a Python type; calling it produces
+        # an instance of the type which is understood by the Teradata driver
+        # in the out parameter mapping mechanism.
+        value = value()
+    return value
```
Not sure that i understand this function

sc250072 replied:
This function is used to translate stored procedure out parameters into a format understandable by the driver. For instance, `str `will be converted to an empty string (''). Stored procedures can be invoked with output parameters in various ways, as illustrated below.
```
TeradataStoredProcedureOperator(
        task_id="opr_sp_types",
        procedure="TEST_PROCEDURE",
        parameters=[3, 1, int, str],
    )
```
This will result in the statement: `{CALL TEST_PROCEDURE(?,?,?,?)}, with parameters: [3, 1, 0, '']`. 
If we omit the usage of this function, the statement would be converted to` {CALL TEST_PROCEDURE(?,?,?,?)}, with parameters: [3, 1, <class 'int'>, <class 'str'>],` which leads to failure with an error.

Similarly, consider another invocation of the `TeradataStoredProcedureOperator`:
```
TeradataStoredProcedureOperator(
        task_id="opr_sp_place_holder",
        procedure="TEST_PROCEDURE",
        parameters=[3, 1, "?", "?"],
 )
```
This will translate to the statement: {CALL TEST_PROCEDURE(?,?,?,?)}, with parameters: [3, 1, ?, ?]. 

Example DAG - https://github.com/apache/airflow/blob/1747e64f51f53a50a62ed31550be9ecf0c5e4ac7/tests/system/providers/teradata/example_teradata_call_sp.py

satish-chinthanippu committed:
pre-commit format applied

Taragolis commented on airflow/providers/teradata/transfers/s3_to_teradata.py:
```diff
@@ -0,0 +1,100 @@
+                                ACCESS_KEY= '{access_secret}'
+                            ) AS d
+                        ) WITH DATA
+                        """
```
I would recommend import `from textwrap import dedent` and remove redundant whitespaces

```suggestion
        sql = dedent(f"""
            CREATE MULTISET TABLE {self.teradata_table} AS
                (
                    SELECT * FROM (
                        LOCATION = '{self.s3_source_key}'
                        ACCESS_ID = '{access_key}'
                        ACCESS_KEY = '{access_secret}'
                ) AS d
            ) WITH DATA
        """).rstrip()
```

sc250072 replied:
Implemented the recommendation.

Taragolis commented on airflow/providers/teradata/transfers/azure_blob_to_teradata.py:
```diff
@@ -0,0 +1,95 @@
+                            ACCESS_KEY= '{access_secret}'
+                    ) AS d
+                    ) WITH DATA
+                """
```
```suggestion
        sql = dedent(f"""
            CREATE MULTISET TABLE {self.teradata_table} AS
                (
                    SELECT * FROM (
                        LOCATION = '{self.blob_source_key}'
                        ACCESS_ID = '{access_id}'
                        ACCESS_KEY = '{access_secret}'
                    ) AS d
                ) WITH DATA
        """).rstrip()
```

sc250072 replied:
Implemented the recommendation.

Taragolis commented on airflow/providers/teradata/transfers/s3_to_teradata.py:
```diff
@@ -0,0 +1,100 @@
+            s3_hook.conn_config.aws_secret_access_key
+            if s3_hook.conn_config.aws_secret_access_key is not None
+            else ""
+        )
```
This one is not semantically correct in case of [AWS Connection](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html#default-connection-ids)

Contain itself might not contain `aws_access_key_id` / `aws_secret_access_key` it could be obtained by [`botocore` / `boto3` credential strategy](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials)

You should call [`get_credentials`](https://github.com/apache/airflow/blob/0a74928894fb57b0160208262ccacad12da23fc7/airflow/providers/amazon/aws/hooks/base_aws.py#L764-L778) method which returns named tuple with frozen credentials

```python
ReadOnlyCredentials = namedtuple(
    'ReadOnlyCredentials', ['access_key', 'secret_key', 'token']
)
```

Please note in case if `token` is not None you should also provide it because without it (STS Session Token) credentials is not valid. 

And I can't find how to do it within the terradata, because manual doesn't contain such information https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/Teradata-VantageTM-Native-Object-Store-Getting-Started-Guide-17.20/Authentication-for-External-Object-Stores/Using-AWS-Assume-Role/Setting-Up-Assume-Role-on-Analytics-Database however [this KB](https://support.teradata.com/knowledge?id=kb_article_view&sys_kb_id=a6b8ad8cc36ae990e2bcb41f050131a7) shows that somehow it supported

And finally there is pretty difficult to handle anonymous access, because there is no out-of-box solution for that in AWS Hooks, so I would recommend to add separate parameter for that, so we could skip obtain connection at all if this kind of access required.

sc250072 replied:
1. Utilizing the [get_credentials](https://github.com/apache/airflow/blob/0a74928894fb57b0160208262ccacad12da23fc7/airflow/providers/amazon/aws/hooks/base_aws.py#L764-L778) function to fetch the access key and secret from the AWS connection.
2. Introduced a new parameter, "public_bucket," to indicate whether the bucket is publicly accessible or not. Accessing get_credentials only when bucket is not public.
3. Concerning the STS session token, we had a JIRA task pending to integrate Teradata authorization object as a parameter for cloud transfer operators. This implementation will be integrated concurrently with the resolution of this JIRA task.

Taragolis replied:
> Concerning the STS session token, we had a JIRA task pending to integrate Teradata authorization object as a parameter for cloud transfer operators.

I would recommend to add into the Operator documentation about current limitation

```

.. note::
   Fo Bar Spam Egg

```

sc250072 replied:
Added documentation as suggested at docs/apache-airflow-providers-teradata/operators/s3_to_teradata.rst https://github.com/apache/airflow/pull/39217/files#diff-763c72396c181b27414fffe4f0cd1a2c97c01868759a119099b56867a03d9e8b

Taragolis commented on tests/system/providers/teradata/example_azure_blob_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                SELECT count(1) from example_blob_teradata_csv;
+            """,
```
There is couple places where triple-quoted string literals redundant not required

```suggestion
        sql="SELECT count(1) from example_blob_teradata_csv;",
```

sc250072 replied:
Addressed suggestion in all applicable places.

Taragolis commented on tests/system/providers/teradata/example_azure_blob_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                DROP TABLE example_blob_teradata_csv;
+            """,
```
```suggestion
        sql="DROP TABLE example_blob_teradata_csv;",
```

sc250072 replied:
Addressed suggestion in all applicable places.

Taragolis commented on tests/system/providers/teradata/example_azure_blob_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                SELECT count(1) from example_blob_teradata_json;
+            """,
```
```suggestion
        sql="SELECT count(1) from example_blob_teradata_json;",
```

Taragolis commented on tests/system/providers/teradata/example_azure_blob_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                DROP TABLE example_blob_teradata_json;
+            """,
```
```suggestion
        sql="DROP TABLE example_blob_teradata_json;",
```

Taragolis commented on tests/system/providers/teradata/example_azure_blob_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                SELECT count(1) from example_blob_teradata_parquet;
+            """,
```
```suggestion
        sql="SELECT count(1) from example_blob_teradata_parquet;",
```

Taragolis commented on tests/system/providers/teradata/example_azure_blob_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                DROP TABLE example_blob_teradata_parquet;
+            """,
```
```suggestion
        sql="DROP TABLE example_blob_teradata_parquet;",
```

Taragolis commented on tests/system/providers/teradata/example_s3_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                    SELECT * from example_s3_teradata_csv;
+                """,
```
```suggestion
        sql="SELECT * from example_s3_teradata_csv;",
```

Taragolis commented on tests/system/providers/teradata/example_s3_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                DROP TABLE example_s3_teradata_csv;
+            """,
```
```suggestion
        sql="DROP TABLE example_s3_teradata_csv;",
```

sc250072 replied:
Addressed suggestion in all applicable places.

Taragolis commented on tests/system/providers/teradata/example_s3_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                    SELECT * from example_s3_teradata_json;
+                """,
```
```suggestion
        sql="SELECT * from example_s3_teradata_json;",
```

sc250072 replied:
Addressed suggestion in all applicable places.

Taragolis commented on tests/system/providers/teradata/example_s3_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                    DROP TABLE example_s3_teradata_json;
+                """,
```
```suggestion
        sql="DROP TABLE example_s3_teradata_json;",
```

Taragolis commented on tests/system/providers/teradata/example_s3_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                    SELECT * from example_s3_teradata_parquet;
+                """,
```
```suggestion
        sql="SELECT * from example_s3_teradata_parquet;",
```

Taragolis commented on tests/system/providers/teradata/example_s3_to_teradata_transfer.py:
```diff
@@ -0,0 +1,163 @@
+        conn_id=CONN_ID,
+        sql="""
+                    DROP TABLE example_s3_teradata_parquet;
+                """,
```
```suggestion
        sql="DROP TABLE example_s3_teradata_parquet;",
```

satish-chinthanippu committed:
Merge branch 'apache:main' into main

satish-chinthanippu committed:
Addressed PR review comments

Addressed following PR review comments

Azure and Amazon optional dependency in teradata provider.yaml
Added another parameter to s3 transfer operator to specify given bucket is public or not
Added more examples of stored procedure operator in stored procedure system DAG
Changed """ to " for single line SQL statements in system tests DAGs
Applied dedent for SQL statements in cloud transfer operators

satish-chinthanippu committed:
unit tests failure issue fixed

Taragolis commented on airflow/providers/teradata/operators/teradata.py:
```diff
@@ -62,3 +66,38 @@ def __init__(
+        self,
+        *,
+        procedure: str,
+        conn_id: str = TeradataHook.default_conn_name,
```
I thought better to use `teradata_conn_id` instead rather than generic `conn_id`

sc250072 replied:
Modified conn_id to teradata_conn_id. It requires to change system tests. Changed system tests as per this modification.

satish-chinthanippu committed:
Address PR comments (#42)

* Modified conn_id to teradata_conn_id for Teradata Operator

* static check issue fixed

potiuk reviewed (approved):
@eladkal @Taragolis ?

satish-chinthanippu committed:
Merge branch 'main' into main

eladkal commented on airflow/providers/teradata/transfers/s3_to_teradata.py:
```diff
@@ -0,0 +1,109 @@
+        The URI format is /s3/YOUR-BUCKET.s3.amazonaws.com/YOUR-BUCKET-NAME.
+        Refer to
+        https://docs.teradata.com/search/documents?query=native+object+store&sort=last_update&virtual-field=title_only&content-lang=en-US
+    :param public_bucket: Specifies whether the provided S3 bucket is public. If the bucket is public,
+        it means that anyone can access the objects within it via a URL without requiring authentication.
+        If the bucket is private and authentication is not provided, the operator will throw an exception.
```
I am not sure about this parameter.
We don't have it in any other transfer to S3.
I would prefer not to add logic on top of what AWS provides.

sc250072 replied:
Teradata has Teradata-supplied public buckets or containers with test data to try out transfers from S3/AzureBlob/Google Storage to teradata. https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/Teradata-VantageTM-Native-Object-Store-Getting-Started-Guide-17.20/Setting-Up-Access/Variable-Substitutions-for-Examples. 
Added public_bucket parameter to specify given S3 path is public. If it is public  bucket/container, the operator skips  getting credentials from S3 Hook https://github.com/apache/airflow/blob/1b1e1cc4905ca993c66b0b8e024d64e9701c2835/airflow/providers/teradata/transfers/s3_to_teradata.py#L89

potiuk replied:
Makes sense to me. @eladkal ?

satish-chinthanippu committed:
Merge branch 'main' into main

satish-chinthanippu committed:
Merge branch 'main' into main

satish-chinthanippu committed:
Merge branch 'main' into main
